test/topology_experimental_raft/test_tablets.py

```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+    servers = dict[ServerNum, ServerInfo]()
+    num_racks = 3
+    nodes_per_rack = 2
+    shards_count = 2
+    initial_tablets = 64
+    rf = num_racks
+    ks = "test"
+    table = "test"
+    for rack in range(1, num_racks + 1):
+        rack_servers = await manager.servers_add(nodes_per_rack, property_file={"dc": "dc1", "rack": f"rack{rack}"})
+        for s in rack_servers:
+            servers[s.server_id] = s
+
+    logger.info("Populate data")
+    cql = manager.get_cql()
+    await cql.run_async(f"CREATE KEYSPACE {ks} WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': {rf}}} AND tablets = {{'initial': {initial_tablets}}};")
+    await cql.run_async(f"CREATE TABLE {ks}.{table} (pk int PRIMARY KEY, c int);")
+
+    keys = range(100)
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO {ks}.{table} (pk, c) VALUES ({k}, 1);") for k in keys])
+
+    logger.info("Verify tablet replicas distribution")
+    tables = {ks: [table]}
+    expected_replicas_per_shard = (initial_tablets * rf) / (num_racks * nodes_per_rack) / shards_count
+    tablet_count = await get_tablet_count_per_shard_for_hosts(shards_count, manager, list(servers.values()), tables)
+    logger.debug(f"Before decommission: expected_replicas_per_shard={expected_replicas_per_shard} tablet_count={tablet_count}")
+    total = 0
+    for server_id, host_replicas in tablet_count.items():
+        s = servers[server_id]
+        for i in host_replicas:
+            total += i
+            assert abs(i - expected_replicas_per_shard) <= 1
+    assert total == initial_tablets * rf
+
+    decommision_rack = f"rack{num_racks}"
+    decommision_count = 0
+    decommisioned_nodes = set[ServerNum]()
+    for s in servers.values():
+        if s.rack == decommision_rack:
+            logger.debug(f"Decommissioning server={s}")
+            decommision_count += 1
+            expected_error = "Unable to find new replica for tablet" if decommision_count == nodes_per_rack else None
+            await manager.decommission_node(s.server_id, expected_error=expected_error)
+            if not expected_error:
+                decommisioned_nodes.add(s.server_id)
+
+    tablet_count = await get_tablet_count_per_shard_for_hosts(shards_count, manager, list(servers.values()), tables)
+    logger.debug(f"After decommission: expected_replicas_per_shard={expected_replicas_per_shard} tablet_count={tablet_count}")
+    total = 0
+    for server_id, host_replicas in tablet_count.items():
+        s = servers[server_id]
+        for i in host_replicas:
+            total += i
+            if s.rack != decommision_rack:
+                assert abs(i - expected_replicas_per_shard) <= 1
+            elif s.server_id in decommisioned_nodes:
+                assert i == 0
+            else:
+                # last remaining node in this rack should contain all replicas in the rack
+                abs(i - expected_replicas_per_shard * nodes_per_rack) <= 1
```

Comment:
no assert here

Reply:
fixed in next version by generalizing verification using a pre-computed map of expected tablets per node